Create spacefish hostname function

Snuggle opened the pull request:
Here's my attempt at converting https://github.com/denysdovhan/spaceship-prompt/blob/master/sections/host.zsh to fish. There may be some mistakes, since I'm quite new to fish. Fixes issue #20.

There are probably issues with this and I'd like to ask for help for integrating + testing. I'm not sure how this is supposed to be done.

<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
I've attempted to convert host.zsh into fish. I'm not quite sure how to integrate/test these changes in Spacefish, however, so there's a high possibility of changes needing to be made.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Attempts to solve #20 and allows people to see their hostname while SSHing into another machine.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
<!-- - [ ] Bug fix (non-breaking change which fixes an issue) --->
- [x] New feature (non-breaking change which adds functionality)
<!--- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected) --->

## Screenshots:
![image](https://user-images.githubusercontent.com/26250962/43232398-801a2346-9068-11e8-89c2-a023077527dd.png)


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
I have run echo statements to ensure the script only runs while SSH is active. I'm not sure how to make the variables available to Spacefish or how Spacefish sources the function. Script shows errors when being run directly, since spacefish::section doesn't appear to exist.
<!--- Include details of your testing environment, tests ran to see how -->
![image](https://user-images.githubusercontent.com/26250962/43171309-667f5720-8fa1-11e8-9a4a-6281ddfa7fcf.png)
Very loosely tested using debug echo statements. Not sure if colors work.
`End` represents the end of the file.
`Is running` represents the main condition that checks if an SSH connection is active, passes.
![image](https://user-images.githubusercontent.com/26250962/43171424-f6510646-8fa1-11e8-8a9e-116e8c5121b6.png)


<!--- your change affects other areas of the code, etc. -->
- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**

I would like to ask for help testing these changes and making sure they integrate correctly into spacefish. Feel free to make any changes. :fish:  

## Default Pull Request Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have checked that no other PR duplicates mine
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly. (Initial attempt made)

## Personal Development Checklist:
- [x] Be able to test my code locally on my system.
- [x] Update Configuration table (Lines 13-20) in `./functions/__sf_section_host.fish` after variable testing locally on my system.
- [x] Add some checks to tests in `./tests/__sf_section_host.test.fish`. (How?)
- [x] Ensure that automated CI testing passes.

Snuggle committed:
Create __sf_section_host.fish

Here's my attempt at converting https://github.com/denysdovhan/spaceship-prompt/blob/master/sections/host.zsh to fish. There may be some mistakes, since I'm quite new to fish. Fixes issue #20.

matchai commented:
Hey Snuggle! Thank you for getting started on #20!
Here are a couple things that will help you get on your way.

[Here](https://github.com/matchai/spacefish/blob/master/fish_prompt.fish#L15) is where the sections are imported.
All sections are prefixed by `__sf_section_` [before they're ran](https://github.com/matchai/spacefish/blob/master/fish_prompt.fish#L30), so you will want to add your section to the list, and rename the file and function to match the formatting.

I would take a look at the following section file to use as reference for code formatting, [setting defaults](https://github.com/matchai/spacefish/blob/master/functions/__sf_section_user.fish#L19-L23) and getting spaceship::section function (`__sf_lib_section` in spacefish) to work: https://github.com/matchai/spacefish/blob/master/functions/__sf_section_user.fish

Looks like you're on the right track! Don't hesistate to ask if you need help with any specifics! 😄
Thanks again for the help!

Snuggle committed:
Fixed formatting and used spacefish utils

This should look a lot closer to https://github.com/matchai/spacefish/blob/master/functions/__sf_section_user.fish

Snuggle committed:
Add host to prompt

Used https://github.com/denysdovhan/spaceship-prompt/blob/master/spaceship.zsh as reference for correct order

Snuggle committed:
Added template configuration table [skip ci]

This will need to be tested and filled out accordingly.

Snuggle committed:
I derped. [skip ci]

Snuggle committed:
Updated documentation.

Snuggle committed:
Created initial test. Some form of checks needed.

Snuggle commented:
@matchai Could you please give me some pointers on how to test my code, both **locally** and with your automated ci-testing? I have experience with Travis, but none with WIP or appveyor.

I need to add some checks to my tests. I'd like to test both in and without an SSH connection, as well as testing that changing the variables functions correctly, but I don't know how to add these tests.

```fish
test "Correctly shows hostname upon SSH connection"
	(
		ssh localhost
		# How to add a check here?
		
	) = (__sf_section_host)
end
```

(Apologies for the three unsigned commits, I derped git signing. :yum:)

Snuggle committed:
Added input to allow SSH to connect

Snuggle committed:
Gah! I broke the tests even more. Will this work?

Snuggle commented:
I'm not sure why the automated tests are failing on a test that I didn't create. :slightly_frowning_face: 

![image](https://user-images.githubusercontent.com/26250962/43196636-8c64a904-9000-11e8-80fa-4de8bc4ebe87.png)

```
test "Don't truncate path"
	(pwd) = (
		__sf_util_truncate_dir (pwd) 0
	)
end
```

Snuggle committed:
Unrelated - Bump fish shell version number [skip ci]

Fish shell 2.7.0 -> Fish shell 2.7.1-

matchai commented:
Looks like you're making a lot of headway! 😄

To run the tests locally, you should be able to simply run `./tests/run.fish` from the project root. That should create a fresh fish and spacefish installation and run the tests. 

Travis and Appveyor should automatically run, and should simply run the same way as it would locally (/tests/run.fish). Travis running on Mac and Linux, and Appveyor running on Cygwin on Windows. WIP simply makes the PR unable to merge until "[WIP]" is removed from its title.

Regarding the random failure; after some tinkering, it looks like it was caused by the unescaped single-quote on [this line](https://github.com/matchai/spacefish/pull/49/files#diff-984f89ec722a751649a7c9c084c65312R14) (not sure why it's getting caught up on a comment). 😅

The tests are ran with [fishtape](https://github.com/fisherman/fishtape), which has *some* documentation, but you might benefit more from referencing some of the pre-existing tests in the repo.

Snuggle commented:
@matchai 

![image](https://user-images.githubusercontent.com/26250962/43232166-84ad0032-9067-11e8-815f-14873011f910.png)
![image](https://user-images.githubusercontent.com/26250962/43232074-1a242790-9067-11e8-8595-a9d11796452c.png)

How is this snippet of code testing the **user** section, though? It's echoing "with snuggle", but my prompt says "in snuggle", how is this test passing? 

"with snuggle" != "snuggle in"

```
➜ ssh localhost
Last login: Wed Jul 25 10:31:39 2018 from 192.168.0.101
Welcome to fish, the friendly interactive shell

snuggle in ~ 
➜ 
```


![image](https://user-images.githubusercontent.com/26250962/43232442-c5eca330-9068-11e8-973a-95ef1a176345.png)
I haven't changed the behavior in my fish config either.

matchai commented:
Good question! 😄

The "with" is the prefix to the user section, while the "in" is the prefix to the dir section.

The prefix of the first section to be printed is hidden by default because of the following config: https://github.com/matchai/spacefish/blob/master/fish_prompt.fish#L10

You can see that in action here:
![image](https://user-images.githubusercontent.com/4658208/43232842-a170bf1e-900e-11e8-88cc-36a34464e593.png)

In the section integration tests, we are running them in isolation, so the prefixes will be shown.

Snuggle committed:
Updated tests, tested locally and all tests should hopefully pass.

Snuggle commented:
This should be complete, hopefully!

I had to remove `SPACEFISH_HOST_SHOW_FULL` as a config option, as it uses Zsh-only functionality that doesn't seem to exist in Fish, but it doesn't seem very useful anyways.

 [Permalink to Zsh's function](https://github.com/denysdovhan/spaceship-prompt/blob/2796da9d5c4a6af7a8255315aa5eeae4aa929679/sections/host.zsh#L34)

I've added ~~six~~ eight tests total, which should test everything within this section. I can only locally test for Windows + Linux, so if you're able to test it'd be appreciated. :slightly_smiling_face: 

It might be worth closing this pull request and waiting on #52.

Snuggle committed:
Added two tests for setting colors.

Snuggle committed:
Attempt to fix travis.ci being weird.

matchai commented on functions/__sf_section_host.fish:
```diff
@@ -0,0 +1,52 @@
+  # Configuration
+  # ------------------------------------------------------------------------------
+
+  # --------------------------------------------------------------------------
```
I appreciate the work put into populating the table!
The table present in `user` was there to match [the one present in spaceship](https://github.com/denysdovhan/spaceship-prompt/blob/master/sections/user.zsh) because of the quirkiness of the user section; so I don't think we'll be needing it here. 🙂

matchai commented on functions/__sf_section_host.fish:
```diff
@@ -0,0 +1,52 @@
+  __sf_util_set_default SPACEFISH_HOST_SHOW true
+  __sf_util_set_default SPACEFISH_HOST_PREFIX "at "
+  __sf_util_set_default SPACEFISH_HOST_SUFFIX $SPACEFISH_PROMPT_DEFAULT_SUFFIX
+  __sf_util_set_default SPACEFISH_HOST_COLOR "blue"
```
Just a heads up to keep things consistent:
For the time being, strings are only wrapped in double-quotes when required (i.e. they contain whitespace, an empty string) 🚀

matchai replied:
I'm not married to this formatting and would be willing to perhaps change it to wrapping all strings in double-quotes. Perhaps a worthwhile decision for a future PR to keep things consistent. 😬

matchai commented on functions/__sf_section_host.fish:
```diff
@@ -0,0 +1,52 @@
+
+  [ "$SPACEFISH_HOST_SHOW" = false ]; and return
+
+  if test $SPACEFISH_HOST_SHOW = 'always'; or set -q SSH_CONNECTION;
```
Apologies for my nitpickiness 😛

In the Fish docs, [it is written](https://fishshell.com/docs/current/commands.html#test):
> When using a variable as an argument for a test operator you should almost always enclose it in double-quotes... Since it is always safe to enclose variables in double-quotes when used as `test` arguments that is the recommended practice.

I would suggest doing that in this case around both the variable and the string. 😁

matchai commented on functions/__sf_section_host.fish:
```diff
@@ -0,0 +1,52 @@
+  [ "$SPACEFISH_HOST_SHOW" = false ]; and return
+
+  if test $SPACEFISH_HOST_SHOW = 'always'; or set -q SSH_CONNECTION;
+    set host_color host
```
In fish, variables are implicitly global unless the `-l` flag is used. 
> -l or --local forces the specified shell variable to be given a scope that is local to the current block, even if a variable with the given name exists and is non-local

As such I would suggest adding it not to clutter the namespace. ✌️

matchai replied:
I don't believe `host` has a value here.
I assume you might be trying to instantiate both variables, but in this case you're assigning the string "host" to `host_color`. 😉

matchai commented on functions/__sf_section_host.fish:
```diff
@@ -0,0 +1,52 @@
+      set host_color $SPACEFISH_HOST_COLOR
+    end
+
+    set host (hostname)
```
Since we're not supporting the `SPACESHIP_HOST_SHOW_FULL` setting, I would consider just using `(hostname)` directly on line 48. 😜

matchai commented on tests/__sf_section_host.test.fish:
```diff
@@ -0,0 +1,130 @@
+test "Display user when SPACEFISH_HOST_SHOW is set to 'always'"
+	(
+		set SPACEFISH_HOST_SHOW always
+		set --erase SSH_CONNECTION
```
To keep the tests atomic, I would suggest putting this line on a new line beneath line 21 as a cleanup step to the last test. 😀

Snuggle replied:
Oh, I tried this and it didn't work since it prematurely unset the SSH_CONNECTION. I think instead of going on line 21 (Green arrow) it has to go after line 22 (Red arrow), since it has to unset SSH *after* the comparison is made.

I don't know why silly me put it in the next test instead after green arrow didn't work.

![image](https://user-images.githubusercontent.com/26250962/43237384-7d9dbcae-9081-11e8-9481-c3030dcf91c5.png)

matchai replied:
Right. That would make sense!

Alternatively, you can create a `teardown` function and clear the `SSH_CONNECTION` variable there.

matchai commented on tests/__sf_section_host.test.fish:
```diff
@@ -0,0 +1,130 @@
+
+test "Display hostname when set different from machine name, over SSH"
+	(
+		function hostname;
```
I add a `mock` function in #51. I'll go ahead and merge that in so you can use it here. 👍

matchai commented on functions/__sf_section_host.fish:
```diff
@@ -0,0 +1,52 @@
+      $SPACEFISH_HOST_PREFIX \
+      $host \
+      $SPACEFISH_HOST_SUFFIX
+      
```
Looks like you dropped a dangling newline!

matchai commented on tests/__sf_section_host.test.fish:
```diff
@@ -0,0 +1,130 @@
+		set_color --bold fff
+		echo -n " "
+		set_color normal
+		
```
Whoops! You left a newline here.

matchai reviewed (approved):
Just a few nitpicky suggestions, but it looks good otherwise!
Thank you your contribution! 🙌

matchai commented:
I would also suggest moving the following line from "upcoming" to "features":
https://github.com/matchai/spacefish/blame/master/README.md#L39

Snuggle committed:
Fix code-consistency issues, update tests

Snuggle committed:
Added suggested changes to README.md

Snuggle committed:
Merge pull request #1 from matchai/master

Add section tests for node (#51)

Snuggle committed:
Merge branch 'master' of github.com:Snuggle/spacefish into patch-1

Snuggle committed:
Use new mock functions!

Snuggle committed:
WHERE'S THE LAMB SOURCE?! 🐑

matchai commented:
Looks squeaky clean to me! 🛁 
We're off to the races!
